### DISCUSSION TITLE
Show image from "raw" bytes (or DynamicImage)

### DISCUSSION THREAD

**Discussion by Bastl34** [emilk/egui#3431] [Q&A] [answered]
How can I show an image which is already present as RGBA-data?
Using the new image api loader seems to just support to load compressed images.

I have tried something like this but without success:
```rust
let pixels = image.as_bytes().to_vec(); // where image is of type DynamicImage (image crate)
let image = egui::Image::from_bytes("bytes://", pixels);
ui.add(image);
```

I'm getting errors that the loader was not found.
I even tried without an uri and with different uri variations.

**Comment by JumpyLionnn** [comment, marked as answer]
First, you need to make sure the image is in the rgba8 format (or rgb). After that you need to convert it to a `ColorImage`.
using the `ColorImage` you can load the texture in the context.
```rust
use image::EncodableLayout;
let color_image = match &image {
      DynamicImage::ImageRgb8(image) => {
          // common case optimization
          egui::ColorImage::from_rgb(
              [image.width() as usize, image.height() as usize],
              image.as_bytes(),
          )
      },
      other => {
          let image = other.to_rgba8();
          egui::ColorImage::from_rgba_unmultiplied(
              [image.width() as usize, image.height() as usize],
              image.as_bytes(),
          )
      },
};
// you must keep the handle, if the handle is destroyed so the texture will be destroyed as well
let handle = ctx.load_texture(name, color_image, options);
let sized_image = egui::load::SizedTexture::new(handle.id(), egui::vec2(color_image.size[0] as f32, color_image.size[1] as f32));
let image = egui::Image::from_texture(sized_image);
```

I hope i helped :)

**Reply by pozix604** [reply]
Is this the correct to do it if I am generating new RGB images every frame? Create handle, load texture, drop the handle, and create a new one next frame?

```rust
let img_data = generate_random_image(128, 128);
let img = ColorImage::from_rgb([128, 128], &img_data);
let handle = ctx.load_texture("noise", img, TextureOptions::default());
let texture = egui::load::SizedTexture::new(handle.id(), egui::vec2(128.0, 128.0));
ui.image(ImageSource::Texture(texture));
```

Not sure if repeated creating new textures every frame is the the intended way.

**Reply by sermuns** [reply]
> Not sure if repeated creating new textures every frame is the the intended way.

I am also wondering this!

**Reply by wangxiaochuTHU** [reply]
> > Not sure if repeated creating new textures every frame is the the intended way.
> 
> I am also wondering this!

Allocate a texture once and then assign a new image to an this existing texture.
```rust
//  texture: &mut TextureHandle  
 texture.set(
    ImageData::Color(colorimage),
    TextureOptions::NEAREST,
  );
```
